executor/runtime/docker/docker.go

```diff
@@ -1873,15 +1873,128 @@ func (r *DockerRuntime) startNonMainContainers(ctx context.Context) error {
 	return group.Wait()
 }
 
-// launchAllContainers starts all existing (pre-created) containers, even the 'main' one (via tini)
-func (r *DockerRuntime) launchAllContainers(ctx context.Context, tiniConns map[string]*net.UnixConn) error {
-	l := log.WithField("taskID", r.c.TaskID())
+func inOrderingAnnotation(thePod *v1.Pod, cName string, suffix string, target string) bool {
+	ann, ok := thePod.Annotations[pod.ContainerAnnotation(cName, suffix)]
+	if ok {
+		for _, c := range strings.Split(ann, ",") {
+			if c == target {
+				return true
+			}
+		}
+	}
+
+	return false
+}
 
+func (r *DockerRuntime) determineContainerOrdering() []string {
 	// First is platform sidecars, then is user-sidecars, then main
 	cNames := r.getPlaformContainerNames()
 	cNames = append(cNames, r.getUserContainerNames()...)
 	cNames = append(cNames, runtimeTypes.MainContainerName)
 
+	origOrder := map[string]int{}
+	for i, cName := range cNames {
+		origOrder[cName] = i
+	}
+
+	thePod, podLock := r.c.Pod()
+	defer podLock.Unlock()
+
+	sort.SliceStable(cNames, func(i, j int) bool {
+		// does i come before j?
+		if inOrderingAnnotation(thePod, cNames[i], pod.AnnotationKeySuffixContainersStartBefore, cNames[j]) {
+			return true
+		}
+
+		// does i come after j?
+		if inOrderingAnnotation(thePod, cNames[i], pod.AnnotationKeySuffixContainersStartAfter, cNames[j]) {
+			return false
+		}
+
+		// does j come before i?
+		if inOrderingAnnotation(thePod, cNames[j], pod.AnnotationKeySuffixContainersStartBefore, cNames[i]) {
+			return false
+		}
+
+		// does j come after i?
+		if inOrderingAnnotation(thePod, cNames[j], pod.AnnotationKeySuffixContainersStartAfter, cNames[i]) {
+			return true
+		}
+
+		// otherwise their order doesn't matter, so let's be stable
+		// w.r.t. the original ordering.
+		return origOrder[cNames[i]] < origOrder[cNames[j]]
+	})
+
+	return cNames
+}
+
+func (r *DockerRuntime) waitForHealthy(ctx context.Context, cName string) error {
+	// XXX: should max tries here? seems like relying on context is enough...
+	for {
+		dockerName := cName
+		if dockerName == runtimeTypes.MainContainerName {
+			dockerName = r.c.TaskID()
```

Comment:
just use container ID consistently here. r.c.ID() is it

See `getContainerNameFromID`